tests/copylat.cpp

```diff
@@ -0,0 +1,273 @@
+                lat_us = ((end.tv_nsec-beg.tv_nsec)/1000.0 + (end.tv_sec-beg.tv_sec)*1000000.0) / (double)iter;
+                printf("cuMemcpy_H2D \t %8zu \t %11.4f\n", copy_size, lat_us);
+                copy_size <<= 1;
+            }
+        } END_CHECK;
+
+        cout << endl;
+        cout << "cuMemcpy_D2H num iters for each size: " << num_read_iters << endl;
+        printf("Test \t\t Size(B) \t Avg.Time(us)\n");
+        BEGIN_CHECK {
+            // cuMemcpy D2H benchmark
+            copy_size = 1;
+            while (copy_size <= size) {
+                int iter = 0;
+                clock_gettime(MYCLOCK, &beg);
+                for (iter = 0; iter < num_read_iters; ++iter) {
+                    ASSERTDRV(cuMemcpy((CUdeviceptr)h_buf, d_A, copy_size));
+                }
+                clock_gettime(MYCLOCK, &end);
+                lat_us = ((end.tv_nsec-beg.tv_nsec)/1000.0 + (end.tv_sec-beg.tv_sec)*1000000.0) / (double)iter;
+                printf("cuMemcpy_D2H \t %8zu \t %11.4f\n", copy_size, lat_us);
+                copy_size <<= 1;
+            }
+        } END_CHECK;
+
+        cout << endl;
+    }
+
+    cout << endl;
+
+    gdr_t g = gdr_open();
+    ASSERT_NEQ(g, (void*)0);
+
+    gdr_mh_t mh;
+    BEGIN_CHECK {
+        // tokens are optional in CUDA 6.0
+        // wave out the test if GPUDirectRDMA is not enabled
+        BREAK_IF_NEQ(gdr_pin_buffer(g, d_A, size, 0, 0, &mh), 0);
+        ASSERT_NEQ(mh, null_mh);
+
+        void *map_d_ptr  = NULL;
+        ASSERT_EQ(gdr_map(g, mh, &map_d_ptr, size), 0);
+        cout << "map_d_ptr: " << map_d_ptr << endl;
+
+        gdr_info_t info;
+        ASSERT_EQ(gdr_get_info(g, mh, &info), 0);
+        cout << "info.va: " << hex << info.va << dec << endl;
+        cout << "info.mapped_size: " << info.mapped_size << endl;
+        cout << "info.page_size: " << info.page_size << endl;
+        cout << "info.mapped: " << info.mapped << endl;
+        cout << "info.wc_mapping: " << info.wc_mapping << endl;
+
+        // remember that mappings start on a 64KB boundary, so let's
+        // calculate the offset from the head of the mapping to the
+        // beginning of the buffer
+        int off = info.va - d_A;
+        cout << "page offset: " << off << endl;
+
+        uint32_t *buf_ptr = (uint32_t *)((char *)map_d_ptr + off);
+        cout << "user-space pointer: " << buf_ptr << endl;
+
+        // gdr_copy_to_mapping benchmark
+        cout << endl;
+        cout << "gdr_copy_to_mapping num iters for each size: " << num_write_iters << endl;
+        printf("Test \t\t\t Size(B) \t Avg.Time(us)\n");
+        copy_size = 1;
+        while (copy_size <= size) {
+            int iter = 0;
+            uint8_t dat;
+            clock_gettime(MYCLOCK, &beg);
+            for (iter = 0; iter < num_write_iters; ++iter) {
+                gdr_copy_to_mapping(mh, buf_ptr, init_buf, copy_size);
+                // ensure that the data reaches the GPU's point of consistency
```

Comment:
this is beyond the aim of this change, as the 3rd party access to device memory is not fully defined at the moment.